### PR TITLE
Validates package executable in Linux

### PR DESCRIPTION
**Pull request by arknable**
Validates package executable in Linux before the package loaded.

### CODE REVIEW

File: cmd/cmd_pkg.go [hunk 1]
```diff
@@ -230,6 +229,24 @@ func loadCommandHandler(cmd *cobra.Command, args []string) {
 		exitWithError(err.Error())
 	}
 
+	// Validate local package
+	packageFolder := filepath.Base(pkgFlags.PackagePath())
```
Comment (eyberg) [inline cmd/cmd_pkg.go:233]: we're really the only ones generating packages today so while this is ok we're mainly concerned with ensuring that one isn't trying to run a macho-o

**Comment by arknable** [comment]
@eyberg it actually works with mach-o
```
Downloading.. https://storage.googleapis.com/nanos/release/nightly/nanos-nightly-linux.tar.gz
 100% |████████████████████████████████████████|  [0s:0s]
package node_v14.2.0 has invalid executable

```

I used [this mach-o binary](https://github.com/JonathanSalwan/binary-samples/blob/master/MachO-OSX-x64-ls) for the test.

**Comment by fabioDMFerreira** [comment]
Commands like `ops run`, `ops image create` and `ops build` also require an elf file to run, so it would be useful to have this validation there. I know #789 is only for validating the elf on local packages though.

To avoid having duplicated code we could place the validation in `cmd/flags_build_image.go`.

File: cmd/cmd_pkg.go [hunk 2]
```diff
@@ -230,6 +229,24 @@ func loadCommandHandler(cmd *cobra.Command, args []string) {
...
+	executableName := c.Program
+	if strings.Contains(executableName, packageFolder) {
+		executableName = filepath.Base(executableName)
+	}
+	executablePath := filepath.Join(pkgFlags.PackagePath(), executableName)
+
+	if runtime.GOOS == "linux" {
```
Comment (fabioDMFerreira) [inline cmd/cmd_pkg.go:240]: Does it make sense to do this for `darwin`?

Comment (eyberg) [reply cmd/cmd_pkg.go:240]: yeh - no need to ensure a platform chk here - the real problem is when someone is building on a mac and forgets to target for linux

**Comment by arknable** [comment]
`run`, `image create`, and `build` already have executable checking and raises following error
`only ELF binaries are supported. Is thia a Linux binary? run "file /home/arknable/.ops/packages/node_v14.2.0/node" on it`

So I will just refactor the validation code as a separated function and call it upon package load

**Comment by fabioDMFerreira** [comment]
@arknable I know what you mean. It's strange that the program is being validated on the method `getSharedLibs`. Anyway, it's a good idea to have the validation code in a dedicated function.

Just make sure you avoid platform-specific code (` runtime.GOOS == "linux"`). Use instead the files `ldd_darwin` and `ldd_linux`. It's ok to create a stub for darwin to make it compile as there is no `isElf` method.

**Comment by arknable** [comment]
> @arknable I know what you mean. It's strange that the program is being validated on the method `getSharedLibs`. Anyway, it's a good idea to have the validation code in a dedicated function.
> 
> Just make sure you avoid platform-specific code (` runtime.GOOS == "linux"`). Use instead the files `ldd_darwin` and `ldd_linux`. It's ok to create a stub for darwin to make it compile as there is no `isElf` method.

@fabioDMFerreira not sure I understand correctly. as we are also checking elf in darwin, I think `isELF` and the validation function should be put inside `ldd.go` so they are available on both os, right ?

**Comment by fabioDMFerreira** [comment]
In the current master branch the `isElf` only exists in `ldd_linux`. I thought the elf validation in darwin had to be done in a different way.

Could you rebase your branch to solve the conflicts? Thanks

**Comment by arknable** [comment]
> In the current master branch the `isElf` only exists in `ldd_linux`. I thought the elf validation in darwin had to be done in a different way.

ah ok, I understand

**Comment by arknable** [comment]
> Could you rebase your branch to solve the conflicts? Thanks

yes I dont know what is the problem but there is no conflict in that file, I already rebased the file against master but the old code keep showing here. I am trying to use github UI to resolve the conflict.

**Comment by arknable** [comment]
> yes I dont know what is the problem but there is no conflict in that file, I already rebased the file against master but the old code keep showing here. I am trying to use github UI to resolve the conflict.

ok it works

**Comment by fabioDMFerreira** [comment]
I've also faced this issue before. I recommend creating a new branch and cherry-pick the commits you want to push.

File: cmd/cmd_pkg.go [hunk 3]
```diff
@@ -228,6 +226,14 @@ func loadCommandHandler(cmd *cobra.Command, args []string) {
 		exitWithError(err.Error())
 	}
 
+	// Validate executable
```
Comment (fabioDMFerreira) [inline cmd/cmd_pkg.go:229]: I think we don't need this comment. The code is pretty clear.

**Comment by arknable** [comment]
> I've also faced this issue before. I recommend creating a new branch and cherry-pick the commits you want to push.

yeah you're right the  web editor merged master to PR's branch. I will move the commits to new branch

**Comment by fabioDMFerreira** [comment]
The build in circleci is failing. If you want to debug locally you can run the script `hack/verify-golint.sh`.

**Comment by arknable** [comment]
> The build in circleci is failing. If you want to debug locally you can run the script `hack/verify-golint.sh`.

ah ok, thanks will do that instead.

**Comment by arknable** [comment]
This PR replaced by https://github.com/nanovms/ops/pull/1032

**Pull request closed without merging**